ForEachAsync Break

MikeBeaton opened the pull request:
This is my attempt to implement my own suggestion #35.

I believe it works and is reasonably sensibly implemented.

Sincere apologies if I've missed an existing way to break a `ForEachAsync` loop without using explicit `throw` and `catch`.

I'd be interested to know what you think.

tyrotoxin commented on Tests/ForEachAsyncTests.cs:
```diff
@@ -31,6 +31,50 @@ public void SimpleSyncForEach()
+                {
+                    Assert.AreEqual(counter, number);
+                    counter++;
+                    if (counter == 3) enumerable.Break();
```
I think this quick solution is a good compromise compared to introducing a loop state argument with the loop control. However, the `Break` extension method is available outside of the loop, and technically should not belong to the enumerable (nor the enumerator).
I suggest to introduce just a simple static class e.g. 'ForEachAsync' with the method `Break`:
```
#if !NETCOREAPP3_0
public static class ForEachAsync
{
    public static void Break() => throw new ForEachAsyncCanceledException();
}
```

So the usage would look like this:
```
await enumerable.ForEachAsync(
item =>
{
    if (...)
        ForEachAsync.Break();
});
```

Thoughts?

MikeBeaton commented:
You seem to have an eye for a good interface, so I've implemented that suggestion!

I've also added some brief documentation.

Also, I may have misunderstood, but it seems that the ForEachAsync.Break is also needed on .NET Core 3.0?

tyrotoxin commented:
Hm, I thought that I removed the `ForEachAsync` extension method in favor of the built-in `foreach-async`. If not, then you are right. Let me review and merge this tonight. Thanks!

MikeBeaton commented:
Hi,

As far as I could make out, in the v3 branch the .NET Core 3 version still
had ForEachAsync, but not publicly exposed? Hope I'm not getting totally
confused!

M

On Mon, 8 Apr 2019, 17:02 Serge Semenov, <notifications@github.com> wrote:

> Hm, I thought that I removed the ForEachAsync extension method in favor
> of the built-in foreach-async. If not, then you are right. Let me review
> and merge this tonight. Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Dasync/AsyncEnumerable/pull/36#issuecomment-480894479>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALZKbQON6ESioNLmnE-BympYeJv7ekeLks5ve2gpgaJpZM4ce4x_>
> .
>
